### PR TITLE
[Concurrency] Store child record when async let child task spawned

### PR DESCRIPTION
**Pull request by ktoso**
Child tasks create a child record and store it when they start.

Without this the new test added here would hang forever, as during cancellation of the parent task we traverse the records in the parent, and cancel all children we find in there. If we didn't record these records it would mean that we would not find the child by traversing the parent on cancellation in swift_task_cancelImpl and they'd never get cancelled.

Group children already worked before, so this is only about async lets.

We also had the idea of all child tasks being task group children, but that's not implemented.
It would mean that every function implicitly becomes a task group, and then children do the exact same as groups do -- this could be quite useful, but that's not yet done in the exact "desugar to exactly an implicit group" way.

Resolves rdar://74464296

Replaces #36645

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: stdlib/public/Concurrency/AsyncLet.swift [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+//@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+//@_silgen_name("swift_asyncLet_create")
+//func _asyncLetCreate(task: Builtin.NativeObject) -> Builtin.RawPointer
+
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+@_silgen_name("swift_asyncLet_start")
+public func _asyncLetStart<T>(
+  asyncLet: Builtin.RawPointer,
+  operation: __owned @Sendable @escaping () async throws -> T
```
Comment (eeckstein) [inline stdlib/public/Concurrency/AsyncLet.swift:27]: The closure is actually not escaping. So the "@escaping" can be removed

Comment (eeckstein) [reply stdlib/public/Concurrency/AsyncLet.swift:27]: Though, we'll need to make sure that the closed over values (the partial_apply arguments) are kept alive until the ``_asyncLetEnd``.

**Comment by ktoso** [comment]
Unrelated failure?

00:32:48.138   Swift(macosx-x86_64) :: ModuleInterface/BadStdlib.swiftinterface

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Also solves rdar://76446550

**Comment by ktoso** [comment]
Just to sanity check we're still good here.

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

**Comment by ktoso** [comment]
Unrelated failure... `00:54:23.248   lldb-api :: lang/swift/async/stepping/step-in/TestSwiftStepInAsync.py`

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS Platform

**Comment by ktoso** [comment]
Disabled the failing test https://github.com/apple/llvm-project/pull/2854

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS Platform

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS

**Comment by ktoso** [comment]
```
01:03:39.169 
/Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift-tools-support-core/Tests/TSCBasicTests/ProcessTests.swift:187: error: -[TSCBasicTests.ProcessTests testSignals] : failed - timeout waiting for signals to be processed
```


😢

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS Platform

**Comment by ktoso** [comment]
@swift-ci please smoke test Windows Platform

**Comment by ktoso** [comment]
@swift-ci please test

Disabled one test on windows.

**Comment by ktoso** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
note to self, the closure is now non escaping thanks to https://github.com/apple/swift/pull/36975

**Pull request merged**